Log KestrelServerOptions when tracing is enabled

davidfowl opened the pull request:
- Dump configured values in a JSON blob and emit an event

```JSON
{
  "AllowSynchronousIO": false,
  "AddServerHeader": true,
  "AllowAlternateSchemes": false,
  "AllowResponseHeaderCompression": true,
  "EnableAltSvc": false,
  "IsDevCertLoaded": true,
  "RequestHeaderEncodingSelector": "default",
  "ResponseHeaderEncodingSelector": "default",
  "Limits": {
    "KeepAliveTimeout": "00:02:10",
    "MaxConcurrentConnections": null,
    "MaxConcurrentUpgradedConnections": null,
    "MaxRequestBodySize": 30000000,
    "MaxRequestBufferSize": 1048576,
    "MaxRequestHeaderCount": 100,
    "MaxRequestHeadersTotalSize": 32768,
    "MaxRequestLineSize": 8192,
    "MaxResponseBufferSize": 65536,
    "MinRequestBodyDataRate": "Bytes per second: 240, Grace Period: 00:00:05",
    "MinResponseDataRate": "Bytes per second: 240, Grace Period: 00:00:05",
    "RequestHeadersTimeout": "00:00:30",
    "Http2": {
      "MaxStreamsPerConnection": 100,
      "HeaderTableSize": 4096,
      "MaxFrameSize": 16384,
      "MaxRequestHeaderFieldSize": 16384,
      "InitialConnectionWindowSize": 131072,
      "InitialStreamWindowSize": 98304,
      "KeepAlivePingDelay": "10675199.02:48:05.4775807",
      "KeepAlivePingTimeout": "00:00:20"
    },
    "Http3": {
      "HeaderTableSize": 0,
      "MaxRequestHeaderFieldSize": 16384
    }
  },
  "ListenOptions": [
    {
      "Address": "https://127.0.0.1:7030",
      "IsTls": true,
      "Protocols": "Http1AndHttp2"
    },
    {
      "Address": "https://[::1]:7030",
      "IsTls": true,
      "Protocols": "Http1AndHttp2"
    },
    {
      "Address": "http://127.0.0.1:5030",
      "IsTls": false,
      "Protocols": "Http1AndHttp2"
    },
    {
      "Address": "http://[::1]:5030",
      "IsTls": false,
      "Protocols": "Http1AndHttp2"
    }
  ]
}
```

The payload size is event source and LOH safe so >= 2K and 4K (with more listen options).

Contributes to #34578

davidfowl committed:
Log KestrelServerOptions when tracing is enabled
- Dump configured values in a dictionary and emit an event with the pairs

davidfowl commented on src/Servers/Kestrel/Core/src/KestrelServerOptions.cs:
```diff
@@ -199,6 +200,20 @@ internal void ApplyDefaultCert(HttpsConnectionAdapterOptions httpsOptions)
+            config[nameof(ResponseHeaderEncodingSelector)] = ResponseHeaderEncodingSelector is null ? "null" : "configured";
+
+            // Limits
+            Limits.Serialize(config);
```
I probably need a "limits." prefix here.

JamesNK commented:
EventSource only allows primitive values afaik. If you want to use eventsource then could could log each name/value in a separate call.

Alternatively, DiagnosticSource supports passing a dictionary.

davidfowl commented:
@noahfalk did some testing and a dictionary worked the other day. But maybe only in narrow cases. I'll likely use JSON if tho doesn't work

davidfowl committed:
Use JSON

davidfowl committed:
Use the Utf8JsonWriter to write configuration values

davidfowl committed:
Added more content
- Added listen options
- Added IsDevCertLoaded property
- Mored RegisterOptions to the end of StartAsync

davidfowl committed:
Remove server options from event source tracking in server stop

davidfowl committed:
Handle new events gracefully

davidfowl commented:
/azp run

davidfowl commented:
/azp run

shirhatti commented:
I'm a little worried about the options going out of sync with the event. Assuming we use this pattern of adding Serialize method everywhere maybe have an analyzer to detect that?

But besides that, looks good to me 👍🏾

Tratcher commented:
IsDevCertLoaded is internal, do we want to log it?

davidfowl commented:
> I'm a little worried about the options going out of sync with the event. Assuming we use this pattern of adding Serialize method everywhere maybe have an analyzer to detect that?

Code reviewz! But a valid concern. We can add a comment to the file but yeah, an analyzer could probably detect it.

> IsDevCertLoaded is internal, do we want to log it?

Seems useful for debugging.

davidfowl commented:
I should probably add a test

JamesNK commented on src/Servers/Kestrel/Core/src/MinDataRate.cs:
```diff
@@ -43,5 +43,10 @@ public MinDataRate(double bytesPerSecond, TimeSpan gracePeriod)
+
+        public override string ToString()
+        {
+            return $"{BytesPerSecond} BPS, Grace Period: {GracePeriod}";
```
nit

```suggestion
            return $"Bytes per second: {BytesPerSecond}, Grace Period: {GracePeriod}";
```

JamesNK commented on src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelEventSource.cs:
```diff
@@ -215,6 +219,54 @@ public void TlsHandshakeFailed(string connectionId)
+        [NonEvent]
+        public void Configuration(KestrelServerOptions options)
+        {
+            var bufferWriter = new ArrayBufferWriter<byte>();
```
Is there an equivalent of checking log level here? Only want to write JSON if Informational and above events are being used.

JamesNK replied:
This:

https://github.com/dotnet/aspnetcore/blob/e470d40c808ca505c8e88bd4678315e6b5cf07b1/src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelEventSource.cs#L63

halter73 replied:
This is checked in AddServerOptions.

noahfalk replied:
The Configuration event is invoked from OnEventCommand and that path ideally would also test the level. If that level test was here it would make it easy to have it enforced uniformly. Currently Configuration() is also a public API so any other code could become a new caller in the future.

halter73 commented:
> "ListenOptions": []

Do we want to log the addresses configured by the `IServerAddressesFeature` when there are no ListenOptions and when `IServerAddressesFeature.PreferHostingUrls` is true? Otherwise this might be a little misleading.

davidfowl commented:
I do but it feels weird to put it on the object. Maybe it's fine?

noahfalk commented on src/Servers/Kestrel/test/InMemory.FunctionalTests/EventSourceTests.cs:
```diff
@@ -423,7 +423,7 @@ public async Task ConnectionLimitExceeded_EmitsStartAndStopEventsWithActivityIds
             Assert.Equal(eventIndex, events.Count);
         }
```
Your choice in deciding how much you care, but if you want to you could write a unit test that deserializes the serialized JSON to see if the value round trips. Another option would be reflecting over all property names to confirm that they match the names observed in the JSON.

davidfowl replied:
I will write a test that JSON deserializes

noahfalk reviewed (approved):
LGTM

davidfowl committed:
Update src/Servers/Kestrel/Core/src/MinDataRate.cs

Co-authored-by: James Newton-King <james@newtonking.com>

davidfowl commented:
> Do we want to log the addresses configured by the IServerAddressesFeature when there are no ListenOptions and when IServerAddressesFeature.PreferHostingUrls is true? Otherwise this might be a little misleading.

Dumped out OptionsInUse (seems to work well).

davidfowl committed:
Print the OptionsInUse
